### PR TITLE
cli `tag`: create a `tag forget` command

### PR DESCRIPTION
**Pull request by ilyagr**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/tag.rs [hunk 1]
```diff
@@ -12,16 +12,22 @@
...
+use crate::cli_util::{parse_string_pattern, user_error, CommandError, CommandHelper};
 use crate::ui::Ui;
 
 /// Manage tags.
 #[derive(clap::Subcommand, Clone, Debug)]
 pub enum TagCommand {
     #[command(visible_alias("l"))]
     List(TagListArgs),
+    #[command(visible_alias("f"))]
+    Forget(TagForgetArgs),
```
Comment (yuja) [inline cli/src/commands/tag.rs:30]: nit: sort them lexicographically so the help is sorted.

File: cli/src/commands/tag.rs [hunk 2]
```diff
@@ -69,3 +88,43 @@ fn cmd_tag_list(
...
+    let mut to_forget = vec![];
+    let mut unmatched_patterns = vec![];
+    for pattern in args.names.iter() {
+        let mut matched = false;
+        for tagname in view.tags().keys().filter(|name| pattern.matches(name)) {
+            matched = true;
+            to_forget.push(tagname.clone());
+        }
+        if !matched {
+            unmatched_patterns.push(pattern.to_string())
```
Comment (yuja) [inline cli/src/commands/tag.rs:110]: nit: `.to_string()` can be omitted

File: cli/src/commands/tag.rs [hunk 3]
```diff
@@ -69,3 +88,43 @@ fn cmd_tag_list(
...
+            "No matching tags for patterns: {}",
+            unmatched_patterns.into_iter().join(", ")
+        )));
+    }
+    for tagname in to_forget.iter() {
+        // Forgetting a tag is the same as deleting it at the moment, as we never push
+        // tags.
+        tx.mut_repo().set_tag_target(tagname, RefTarget::absent());
+    }
+    tx.finish(ui, format!("forgot {} tags", to_forget.len()))?;
```
Comment (yuja) [inline cli/src/commands/tag.rs:124]: nit: `forget` (present tense)

and maybe better to include names `to_forget.iter().join(", ")`.

File: cli/src/commands/tag.rs [hunk 4]
```diff
@@ -69,3 +88,43 @@ fn cmd_tag_list(
...
+
+    if to_forget.len() > 1 {
+        writeln!(ui.stderr(), "Forgot {} branches.", to_forget.len())?;
```
Comment (yuja) [inline cli/src/commands/tag.rs:127]: nit: `tags`

File: cli/tests/test_tag_command.rs [hunk 5]
```diff
@@ -77,3 +79,70 @@ fn test_tag_list() {
...
+#[test]
+fn test_tag_forget() {
+    let test_env = TestEnvironment::default();
+    test_env.add_config("git.auto-local-branch = true");
+    let git_repo_path = test_env.env_root().join("source");
+    let git_repo = git2::Repository::init(git_repo_path).unwrap();
+
+    set_up_tagged_git_repo(&git_repo);
+
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "clone", "source", "tagged"]);
```
Comment (yuja) [inline cli/tests/test_tag_command.rs:92]: can you add colocated test?

I expect it wouldn't work because forgotten tags aren't exported.

Comment (ilyagr) [reply cli/tests/test_tag_command.rs:92]: Good point! I guess I should see about exporting tag deletion in this case.

Comment (yuja) [reply cli/tests/test_tag_command.rs:92]: > that might cause tag deletion on a git push

You mean `git push` (not `jj git push`)?

We don't have `remote_views["git"].tags` yet, but I think exporting forgotten tags is okay.

Comment (ilyagr) [reply cli/tests/test_tag_command.rs:92]: Yes, I mean the Git `git push --tags`. There is no problem if we're talking about tags that are already deleted on the remote, but I'm worried that if they aren't, Git will delete them on the remote on push.

We could try to delete remote-tracking Git tags, if there is such a thing.

If there is such a think as remote-tracking tags, another alternative would be to only allow forgetting a tag if no "remote-tracking tags" exist in Git.

Every way seems messy.

Comment (yuja) [reply cli/tests/test_tag_command.rs:92]: Tag handling in Git is peculiar. I don't remember how to delete remote tags by git, but we'll need `--delete` or something?
https://git-scm.com/book/en/v2/Git-Basics-Tagging#_deleting_tags

**Comment by ilyagr** [comment]
The issue at https://github.com/martinvonz/jj/pull/3097#discussion_r1495692790 seems fatal; we'd have to implement `tag@git` refs by analogy to `branch@git` IIRC.

**Pull request closed without merging**